### PR TITLE
refactor(api): restrict arbitrary input nesting

### PR DESCRIPTION
**Pull request by kszucs**
Previously we allowed arbitrary nesting of input expressions for various
API methods like `table.join`, `table.group_by`, etc. While this was
backward compatible with `8.0.0` it also exposes additional ambiguity
which can be confusing for users and difficult to reason about.

This change restricts the nesting of input expressions to a
"single level" of nesting with the exception of the first positional
argument which can be a list of expressions, but only if there are no
more positional arguments. Examples:

```python
t.select([t.foo, t.bar])  # OK
t.select(t.foo, t.bar)    # OK
t.select([t.foo], t.bar)  # Error
t.select(t.foo, name=t.bar)  # OK
t.select([t.foo], name=t.bar)  # OK
```

Aims to resolve https://github.com/ibis-project/ibis/issues/7819 and https://github.com/ibis-project/ibis/issues/8304, depends on https://github.com/ibis-project/ibis/pull/8884

### CODE REVIEW

File: ibis/expr/types/relations.py
Comment (kszucs) [inline ibis/expr/types/relations.py:284]: Needs docstrings and must be covered with tests, but first we should have a consensus on the API behaviour.

**Comment by kszucs** [comment]
There are some ambiguities around the first arguments which can be a list of expressions:

```py
table.select(["a", ["b"]])

# constructs

r0 := UnboundTable: table
  a int8
  b int16
  c int32
  d int64
  e float32
  f float64
  g string
  h boolean
  i timestamp
  j date
  k time

Project[r0]
  a:      r0.a
  ('b',): ['b']
```

`a` is interepreted as a `column` whereas `["b"]` interpreted as an `array literal`.

**Comment by kszucs** [comment]
Apparently we also have a backend test case using a mapping as a single positional argument.

**Comment by cpcloud** [comment]
Why can't we keep

```python
t.select([t.foo], t.bar)
```
?

It doesn't seem meaningfully different from allowing

```python
t.select([t.foo], bar=t.bar)
```

**Comment by kszucs** [comment]
If we allow:

```py
t.select([t.foo], t.bar)
```

then should we allow

```py
t.select([t.foo], [t.bar])
```

as well?

**Comment by cpcloud** [comment]
I think so. It's really things like `t.select([[...]], [[...]])` that should be disallowed.

**Comment by kszucs** [comment]
My preference would be to either provide args as variadics or as a single list-like argument. Even then we have confusion when to treat list-like inputs as literals vs value list.

**Comment by jcrist** [comment]
We talked today during triage and decided that at minimum we want to remove the support for arbitrarily nested expressions (as Phillip noted above).

I actually agree with the initial proposal at the top (only accept lists of column refs if select is a 1-arg function, otherwise treat it as variadic). Coincidentally this matches how polars `select` works, so there's some ecosystem-consistency there. The trick with doing this as a breaking change though is that previously working code may instead silently be treated as an array literal.

With this PR, what happens here? Does this error, or does it silently coerce these to array literals?

```python
t.select(["a", "b"], ["c", "d"])  # is this two array literals or an error?

t.select("a", "b", ["c", "d"])   # is ["c", "d"] an array literal, or does this error?
```

Either way, I think we _definitely_ want to forbid the case that phillip brought up before 9.0. I'm not opposed to the full changes here either, I think they make things more consistent, but we can always do that later IMO in 10.0.

**Comment by kszucs** [comment]
```py
t.select(["a", "b"], ["c", "d"])  # is this two array literals or an error?
```
yes, two array literals, not erroring

```py
t.select("a", "b", ["c", "d"])   # is ["c", "d"] an array literal, or does this error?
```
yes, an array literal, not erroring

**Comment by kszucs** [comment]
I would prefer deciding what should be the ideal long term behaviour then extend that to be backward compatible and either raise a deprecation warning or provide a config option to keep the previous behaviour.

**Comment by cpcloud** [comment]
Let's keep the full changeset here, and make sure to have both a `BREAKING CHANGE` bit as well as documentation on the behavior in the relevant APIs (`select` and `mutate`).

**Comment by kszucs** [comment]
Another thing I'm not sure about is whether we should do the dereferencing here or not? Like should `t.bind(totally_different_table.col)` raise?

**Comment by cpcloud** [comment]
Seems like it _might_ keep `bind` more flexible and easier to maintain if we dereference elsewhere.

**Comment by kszucs** [comment]
Actually it could help code consolidation. I can try it after #8916 gets merged.

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [inline ibis/tests/expr/test_table.py:2086]: Would love to see test cases for every ifelse statement in the implementation:
- bools (raw, no ibis.literal(bool))
- ints (raw)
- lambda
- a Table

**Review by NickCrews — CHANGES_REQUESTED** [review]

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [reply ibis/tests/expr/test_table.py:2081]: Also a test for deferreds, eg _.bind("a")

**Comment by cpcloud** [comment]
We should probably kill `destructure()` now that proper alternatives are available.

**Comment by cpcloud** [comment]
I'm going to rip out `destructure` in a separate PR.

File: ibis/expr/types/relations.py
Comment (NickCrews) [reply ibis/expr/types/relations.py:94]: that workflow makese sense, thanks! Indeed these requests aren't hard blockers. I think my request for test cases is a blocker for me though, see my comment down there.

File: ibis/tests/expr/test_table.py
Comment (kszucs) [reply ibis/tests/expr/test_table.py:2081]: Table was already there, added tests for the rest.
`bind(_, "column_name")` case isn't supposed to work since `bind()` expects an actual table expression not a deferred as the first argument. Adding a test case for `bind(t, _.column_name)`.

**Comment by kszucs** [comment]
> I'm going to rip out destructure in a separate PR. 

Updated the test case to expect an input error rather than an integrity error. We should update `unwrap_aliases()` to raise ibis input error instead of integrity error too.

File: ibis/tests/expr/test_table.py
Comment (kszucs) [inline ibis/tests/expr/test_table.py:276]: Enable this test case.

File: ibis/tests/expr/test_table.py
Comment (kszucs) [inline ibis/tests/expr/test_table.py:271]: Remove this test case.

File: ibis/tests/expr/test_table.py
Comment (kszucs) [reply ibis/tests/expr/test_table.py]: Done.

**Comment by kszucs** [comment]
One confusing case:

```py
In [4]: table = ibis.table({"a": "int", "b": "string"})

In [5]: table.select(["a", ["b"]])
Out[5]:
r0 := UnboundTable: unbound_table_1
  a int64
  b string

Project[r0]
  a:      r0.a
  ('b',): ['b']
```

File: ibis/expr/types/relations.py
Comment (kszucs) [reply ibis/expr/types/relations.py:94]: `t.bind()` also dereference the value to `t` now. `bind()` simply coerces the input to a value expression and remains a private function.

File: ibis/expr/types/relations.py
Comment (kszucs) [reply ibis/expr/types/relations.py:213]: Tests are there but the docstring is missing.

File: ibis/expr/builders.py
Comment (cpcloud) [inline ibis/expr/builders.py:229]:
```suggestion
```

**Review by cpcloud — APPROVED** [review]

**Comment by cpcloud** [comment]
@NickCrews Can you review this again? This is a remaining blocker for the 9.0 release which we'd like to get out early next week.

File: ibis/expr/types/relations.py
Comment (NickCrews) [inline ibis/expr/types/relations.py:773]: Am I understanding this right that `.__getitem__` accepts ints, but `.select` doesn't? This is a separate PR, but I would love to rip the bandaid off and just disallow ints everywhere.

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [inline ibis/tests/expr/test_table.py:2119]: based on the comment I think this was supposed to be a tuple, but its a list. But maybe go way weird and pass in an iterable to really stress it: `(x for x in [1, "a"])`

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [inline ibis/tests/expr/test_table.py:2094]: can we get a test for when the resulting columns would have duplicate names? eg `t.bind(1, 1)`

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [inline ibis/tests/expr/test_table.py:2087]: would love a case for no args eg `t.bind()`. I would expect this to result in an empty tuple. Erroring would be OK if there is some problem with the empty tuple, but that would be my second choice.

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [inline ibis/tests/expr/test_table.py:2092]: Would love 2 cases, one for for a completely unrelated table `ibis.table({"z": "int"}).bind(t)` and another for a related one (I can only think of a join result but maybe there's some other instance of this) `t.right_join(ibis.table({"a": "int"}), "a").bind(t.b)`. I'm not that attached to what the behavior is, but I think we should write it down so we can test for future breakages.

File: ibis/expr/types/relations.py
Comment (cpcloud) [reply ibis/expr/types/relations.py:773]: I think it's still useful to be able to do this. I use it during ad-hoc interactive analysis for referring to columns that have generated names.

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [reply ibis/tests/expr/test_table.py:2092]: Don't we already have a test for the latter?

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [reply ibis/tests/expr/test_table.py:2119]: This doesn't seem like a blocker to me.

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [reply ibis/tests/expr/test_table.py:2099]: This doesn't seem like a blocker.

File: ibis/expr/types/relations.py
Comment (NickCrews) [inline ibis/expr/types/relations.py:215]: I'm a -.5 on this special casing. This causes `t.bind([1,"a"])`  to be `(ibis.literal(1), t.a)` (behavior A). I would prefer for it to be `ibis.array([1, "a"])` (behavior B). I looked back through this PR to try to get the context, but I might not have gotten all of it. As I understand it, switching A to B

- Would be the preferred behavior by everyone else if we were starting from a clean slate with no history.
- Might silently change the behavior of existing user code (we could add an easily-silencable warning for the 9.0 and maybe 10.0 versions?). This is the main downside.
- Would this cause any other paper cuts or make impossible some other task I'm not thinking of?

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [reply ibis/tests/expr/test_table.py:2099]: Ok, this isn't actually supported. I'll push up a fix.

Comment (cpcloud) [reply ibis/tests/expr/test_table.py]: Done.

File: ibis/expr/types/relations.py
Comment (NickCrews) [inline ibis/expr/types/relations.py:234]: This would gobble up unrelated errors: `t.bind(lambda t: int("foo"))`. Let's be a little slower here:
```
bindings = bind(self, arg)
if len(bindings) != 1:
    raise com.IbisInputError(
                    "Keyword arguments cannot produce more than one value"
                )
values.append(bindings[0].name(key))
```

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [reply ibis/tests/expr/test_table.py]: Added a test with a generator.

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [reply ibis/tests/expr/test_table.py:2092]: do you mean the `t.bind(t)` case? I was hoping for something a bit more complex. Or did I miss something?

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [reply ibis/tests/expr/test_table.py:2094]: I see what you're saying, but I think I still want it. Perhaps in the future we switch to a different implementation, would still want the test.

File: ibis/expr/types/relations.py
Comment (NickCrews) [reply ibis/expr/types/relations.py:775]: yeah I see what you're saying. I was trying to get these two methods to be totally equivalent, but I think I am convinced that the utility is worth the inconsistency.

**Comment by NickCrews** [comment]
I just took a long look.

Most things look great, a few nits, but the one substantial thing is the `t.bind([1,"a"])` behavior. I have an inline comment above. Can someone fill me in on the context there a little more? I don't think this is a blocker for me.

File: ibis/tests/expr/test_table.py
Comment (cpcloud) [reply ibis/tests/expr/test_table.py:2094]: Let's separate blocking comments from non blocking comments.

Adding a test for this should in no way block this PR from landing.

I disagree that we should add this test. It adds no material value to testing the behavior of `bind`. Adding a test that exercises duplicate column detection is just wasted CPU cycles and maintainer effort.

**Comment by cpcloud** [comment]
@NickCrews Thanks for doing a deep dive on tests. There were indeed a few cases that weren't tested and also the `lambda _: 2` case wasn't yet implemented. Appreciate the feedback!

File: ibis/expr/types/relations.py
Comment (cpcloud) [reply ibis/expr/types/relations.py:217]: > Might silently change the behavior of existing user code (we could add an easily-silencable warning for the 9.0 and maybe 10.0 versions?).

The benefit of behavior B doesn't seem worth the pain this would cause given history.

I don't think changing to behavior B makes anything impossible, but it's a breaking API change that seems unnecessary.

File: ibis/expr/types/relations.py
Comment (cpcloud) [reply ibis/expr/types/relations.py:234]: Yeah, we should avoid catching the error here. I'll fix it up.

Comment (cpcloud) [reply ibis/expr/types/relations.py]: Done and test added.

File: ibis/expr/types/relations.py
Comment (kszucs) [reply ibis/expr/types/relations.py:217]: > Would be the preferred behavior by everyone else if we were starting from a clean slate with no history.

It would be my preference, but yes, that would be too big of a breakage. If we wish to change it we can also do that for later releases.

File: ibis/expr/types/relations.py
Comment (kszucs) [reply ibis/expr/types/relations.py]: Nice catch! Thanks for the patch too!

File: ibis/tests/expr/test_table.py
Comment (kszucs) [reply ibis/tests/expr/test_table.py:2094]: So bind's responsibility to coerce the input arguments to a series of value expressions, nothing more.

Other methods like `.select() / .aggregate() / ...` do ensure that there are no duplicate names by calling `unwrap_aliases()`, so the check does happen but at a different level.

**Comment by NickCrews** [comment]
Cool this looks great then!

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [reply ibis/tests/expr/test_table.py:2092]: I think it is a blocker for me, at the very least to simply raise a NotImplemented error so we keep the door open for ourselves later

File: ibis/expr/types/relations.py
Comment (NickCrews) [reply ibis/expr/types/relations.py:217]: sounds good let's keep it as is for now. Thank you!

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [reply ibis/tests/expr/test_table.py:2092]: oooh, actually there is also `t.bind(unrelated_table.column)` which currently returns `unrelated_table.column`, which I don't think quite makes sense (I would expect an error unless we can walk up the join chain to find it). So actually all this seems like undefined behavior. I think if we just don't publicize this API or add it to the type hints then I'm happy. Would love to make this more flushed out later but it is a bigger task than we should block on.

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [reply ibis/tests/expr/test_table.py:2094]: OK, I think this is a misunderstanding caused by me not explaining what I thought the expected behavior should be. I have always used .bind in the larger context of `t.mutate(t.bind(x))`, so I have never wanted duplicate column names, so I originally thought that `t.bind(1, 1)` should error. But

> So bind's responsibility to coerce the input arguments to a series of value expressions, nothing more.

makes sense. I think you're right, dupe results should be allowed, and treated the same as any other result, and thus don't need a test. I was operating under the assumption they should be treated differently, and thus should have a test.

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [reply ibis/tests/expr/test_table.py:2092]: Or I guess per the conversation below, maybe again this isn't the responsibility of .bind to do this check

File: ibis/tests/expr/test_table.py
Comment (NickCrews) [reply ibis/tests/expr/test_table.py:2094]: I would maybe tweak the phrasing of the responsibility to "coerce the arguments to a tuple of value expressions (if they aren't already)." to emphasize that we don't do anything more clever/furhter inspection if the arg is already a value expression. This makes it more clear that `t.bind(other_table.column)` just results in `other_table.column`. I think this would be a good candidate to go in the summary line of the docstring.

**Comment by cpcloud** [comment]
@NickCrews Can you approve the PR?

**Review by NickCrews — APPROVED** [review]

**Comment by NickCrews** [comment]
Sorry I was on the mobile app where the approval status isn't even shown so i didn't know there was anything to do

**Pull request merged**
